models/iaf_wang_2002.cpp

```diff
@@ -0,0 +1,541 @@
+
+    while ( t < B_.step_ )
+    {
+      const int status = gsl_odeiv_evolve_apply( B_.e_,
+        B_.c_,
+        B_.s_,
+        &B_.sys_,              // system of ODE
+        &t,                    // from t
+        B_.step_,              // to t <= step
+        &B_.integration_step_, // integration step size
+        S_.y_ );       // neuronal state
+
+      if ( status != GSL_SUCCESS )
+      {
+        throw GSLSolverFailure( get_name(), status );
+      }
+    }
+
+
+    // add incoming spikes
+    S_.y_[ State_::s_AMPA ] += B_.spike_AMPA.get_value( lag );
+    S_.y_[ State_::s_AMPA_ext ] += B_.spike_AMPA_ext.get_value( lag );
+    S_.y_[ State_::s_GABA ] += B_.spike_GABA.get_value( lag );
+    S_.y_[ State_::s_NMDA ] += B_.spike_NMDA.get_value( lag );
+    if ( S_.y_[ State_::s_NMDA ] > 1 )
+    {
+      S_.y_[ State_::s_NMDA ] = 1;
+    }
+    if ( S_.r_ )
+    {
+      // neuron is absolute refractory
+      --S_.r_;
+      S_.y_[ State_::V_m ] = P_.V_reset; // clamp potential
+    }
+    else if ( S_.y_[ State_::V_m ] >= P_.V_th )
+    {
+      // neuron is not absolute refractory
+      S_.r_ = V_.RefractoryCounts_;
+      S_.y_[ State_::V_m ] = P_.V_reset;
+
+      // get previous spike time
+      double t_lastspike = get_spiketime_ms();               
+
+      // log spike with ArchivingNode
+      set_spiketime( Time::step( origin.get_steps() + lag + 1 ) );
+
+      double t_spike = get_spiketime_ms();
+
+      // compute current value of s_NMDA and add NMDA update to spike offset
+      S_.s_NMDA_pre = S_.s_NMDA_pre * exp( -( t_spike - t_lastspike ) / P_.tau_decay_NMDA );
+      double s_NMDA_delta = P_.alpha * (1 - S_.s_NMDA_pre);
+      S_.s_NMDA_pre += s_NMDA_delta;
+
+      SpikeEvent se;
+      se.set_offset( s_NMDA_delta );
+      kernel().event_delivery_manager.send( *this, se, lag );
+    }
+
+    // set new input current
+    B_.I_stim_ = B_.currents_.get_value( lag );
+
+    // voltage logging
+    B_.logger_.record_data( origin.get_steps() + lag );
+  }
+}
+
+// Do not move this function as inline to h-file. It depends on
+// universal_data_logger_impl.h being included here.
+void
+nest::iaf_wang_2002::handle( DataLoggingRequest& e )
+{
+  B_.logger_.handle( e );
+}
+
+void
+nest::iaf_wang_2002::handle( SpikeEvent& e )
+{
+  assert( e.get_delay_steps() > 0 );
+
+  if ( e.get_weight() > 0.0 )
+  {
+    if ( e.get_rport() == 0 ) { // recurrent spike
+      B_.spike_AMPA.add_value( e.get_rel_delivery_steps( kernel().simulation_manager.get_slice_origin() ),
+        e.get_weight() * e.get_multiplicity() );
+      B_.spike_NMDA.add_value( e.get_rel_delivery_steps( kernel().simulation_manager.get_slice_origin() ),
+        e.get_weight() * e.get_multiplicity() * e.get_offset() );
+    }
+    else if ( e.get_rport() == 1 ) { // external spike
+      B_.spike_AMPA_ext.add_value( e.get_rel_delivery_steps( kernel().simulation_manager.get_slice_origin() ),
+        e.get_weight() * e.get_multiplicity() );
```

Comment:
Why this, what is an "external spike"?

Since offset is 0 for spikes sent from other neuron models (except precise spiking, might need a check here), we could run all spikes through rport 0.